[iOS] Tapping on Now Playing opens wrong app when using SFSafariViewController

jernoble opened the pull request:
#### 084e01d07525c8d55324a76d903c0ee5dd12b64b
<pre>
[iOS] Tapping on Now Playing opens wrong app when using SFSafariViewController
<a href="https://bugs.webkit.org/show_bug.cgi?id=243481">https://bugs.webkit.org/show_bug.cgi?id=243481</a>

Reviewed by Chris Dumez.

When creating a NowPlayingInfo struct, we use the sourceApplicationIdentifier,
which when running via SFSafariViewController, will be &quot;com.apple.SafariViewService&quot;.
But NowPlaying wants to know the presenting application, so that bundle identifier
must be plumbed down from the UIProcess to the WebContent process. The presenting
application PID is already sent, but is insufficient for the API we need to call.

* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::nowPlayingInfo const):
* Source/WebCore/platform/RuntimeApplicationChecks.h:
* Source/WebCore/platform/cocoa/RuntimeApplicationChecksCocoa.mm:
(WebCore::presentingApplicationBundleIdentifierStorage):
(WebCore::setPresentingApplicationBundleIdentifier):
(WebCore::presentingApplicationBundleIdentifier):
* Source/WebKit/Shared/WebProcessCreationParameters.cpp:
(WebKit::WebProcessCreationParameters::encode const):
(WebKit::WebProcessCreationParameters::decode):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: <a href="https://commits.webkit.org/253115@main">https://commits.webkit.org/253115@main</a>
</pre>

cdumez commented on Source/WebCore/platform/cocoa/RuntimeApplicationChecksCocoa.mm:
```diff
@@ -91,6 +91,22 @@ void clearApplicationBundleIdentifierTestingOverride()
 
+static String& presentingApplicationBundleIdentifierStorage()
+{
+    static NeverDestroyed<String> identifier;
```
We likely want MainThreadNeverDestroyed here given that this doesn't look thread safe.

jernoble replied:
I'll also add a `ASSERT(RunLoop::isMain())`, to verify this is only being accessed on the main thread.

jernoble replied:
Aha, never mind. It looks like MainThreadNeverDestroyed already does that.

cdumez replied:
Right, that is the whole point of MainThreadNeverDestroyed AFAIK :)

cdumez commented on Source/WebCore/platform/RuntimeApplicationChecks.h:
```diff
@@ -67,6 +67,9 @@ WEBCORE_EXPORT void setApplicationBundleIdentifierOverride(const String&);
 WEBCORE_EXPORT void clearApplicationBundleIdentifierTestingOverride();
 
+WEBCORE_EXPORT void setPresentingApplicationBundleIdentifier(const String&);
+WEBCORE_EXPORT String presentingApplicationBundleIdentifier();
```
Should probably return a `const String&`

jernoble committed:
[iOS] Tapping on Now Playing opens wrong app when using SFSafariViewController
https://bugs.webkit.org/show_bug.cgi?id=243481

Reviewed by Chris Dumez.

When creating a NowPlayingInfo struct, we use the sourceApplicationIdentifier,
which when running via SFSafariViewController, will be "com.apple.SafariViewService".
But NowPlaying wants to know the presenting application, so that bundle identifier
must be plumbed down from the UIProcess to the WebContent process. The presenting
application PID is already sent, but is insufficient for the API we need to call.

* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::nowPlayingInfo const):
* Source/WebCore/platform/RuntimeApplicationChecks.h:
* Source/WebCore/platform/cocoa/RuntimeApplicationChecksCocoa.mm:
(WebCore::presentingApplicationBundleIdentifierStorage):
(WebCore::setPresentingApplicationBundleIdentifier):
(WebCore::presentingApplicationBundleIdentifier):
* Source/WebKit/Shared/WebProcessCreationParameters.cpp:
(WebKit::WebProcessCreationParameters::encode const):
(WebKit::WebProcessCreationParameters::decode):
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: https://commits.webkit.org/253115@main
